### PR TITLE
Add complex subgrouping support

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

File: implementation.md [hunk 1]
```diff
@@ -162,6 +162,27 @@ the JSON representation and `ttl` indicating the number of minutes to cache the
 input. Additionally, once a Shesmu server is active, it will provide the input
 in the JSON format at `/input/` followed by the input format name.
 
+## Writing a Grouper Plugin
+Creating a new grouper is relatively straight forward, but thorny since a lot
```
Comment (callunity) [inline implementation.md]: `relatively straightforward` <- I'm not sure I agree... :stuck_out_tongue:

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/BaseOliveBuilder.java [hunk 2]
```diff
@@ -364,6 +373,167 @@ public final void measureFlow(String filename, int line, int column) {
...
+        new LambdaBuilder(
+            owner,
+            String.format("Group with Grouper %d:%d ✨", line, column),
+            LambdaBuilder.function(newType, oldType),
+            capturedVariables);
+
+    // Since every grouper can apply extra arguments, we need to prepare a lambda that takes
+    // those extra arguments and captures them into a final BiConsumer that the wrapper can use to
+    // collect the values. This is that capturing method. It first captures all the external values
+    // into a type specified by the grouper. Inside it creates a new lambda capturing the eternal
```
Comment (callunity) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/BaseOliveBuilder.java]: `eternal` -> `external` ?

File: language.md [hunk 3]
```diff
@@ -156,12 +161,49 @@ that in the `By` clause:
...
+      By workflow
+      Using always_include # This will be the grouper name
+				# These are grouper-specific parameters that control how the grouping
+				# works
+        key = metatype,
+        include_when = "inode/directory"
+			# The grouper will not only create groups but can provide additional
+			# information about the groups. This additional information will be
+			# available with the names `current_metatype` and `is_directory`. The
+			# `With` can be omitted and the default names will be used.
```
[the remark this answers is not in the document]
Comment (apmasell) [reply language.md]: Not defaults, those are required. The `With` has defaults.

File: implementation.md [hunk 4]
```diff
@@ -162,6 +162,27 @@ the JSON representation and `ttl` indicating the number of minutes to cache the
...
+rows. These are opaque to the grouper and Shesmu will fill in the gaps using
+the olive.
+
+1. Create a class _G_ that implements `Grouper` parameterised over _I_ and _O_.
+1. Create a class _D_ that extends `GrouperDefinition` and is annotated with
+`@MetaInfServices`.
+1. In _D_, call the appropriate super constructor. They vary by the number of
+input parameters the grouper requires. Use `GrouperParameter` to fill in each
+parameter required.
+1. In _D_, select a `GrouperOuputs` to use. This set the number of variables
```
[the remark this answers is not in the document]
Comment (apmasell) [reply implementation.md]: `GrouperOutputs` is the name of a class.

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/GrouperDefinition.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+              public Grouper<I, O> pack(
+                  Object first, Object second, Object third, C outputGenerator) {
+                return factory.pack(
+                    parameter1.unpack(first),
+                    parameter2.unpack(second),
+                    parameter3.unpack(third),
+                    outputGenerator);
+              }
+            }));
+    this.inputs = Arrays.asList(parameter1, parameter2);
```
[the remark this answers is not in the document]
Comment (apmasell) [reply shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/GrouperDefinition.java]: Yes. Yikes.

File: implementation.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (callunity) [reply implementation.md]: Sorry, I added an extra space in my suggestion. `Outputs` is misspelled here.

File: language.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (callunity) [reply language.md]: Then perhaps `existing keys` as opposed to `default names`? It's not clear what `the default names` refers to.

[the remark this answers is not in the document]
Comment (apmasell) [reply language.md]: I've tried to rewrite this.

File: language.md [hunk 6]
```diff
@@ -156,12 +161,51 @@ that in the `By` clause:
...
+          files = List path
       Run fingerprint With
         memory = 4Gi,
         input = files;
 
+The grouping shown so far requires that the groups being produced are known
+ahead of time. In some situations, it isn't possible to know exactly which item
+belong in which groups until all the data is available. For these situations,
+_groupers_ are available that can do complex subgrouping. The groupers are
+plugins, so the groupers available can been seen on the running Shemsu server.
```
Comment (iainrb) [inline language.md]: "can been seen" -> "can be seen"

File: language.md [hunk 7]
```diff
@@ -156,12 +161,51 @@ that in the `By` clause:
...
-          files = List path
         By project, sequencer_run = ius[0]
+        Into
+          files = List path
       Run fingerprint With
         memory = 4Gi,
         input = files;
 
+The grouping shown so far requires that the groups being produced are known
+ahead of time. In some situations, it isn't possible to know exactly which item
```
Comment (iainrb) [inline language.md]: "which item belong" -> "which items belong"

File: language.md [hunk 8]
```diff
@@ -156,12 +161,51 @@ that in the `By` clause:
...
+      By workflow
+      Using always_include # This will be the grouper name
+				# These are grouper-specific parameters that control how the grouping
+				# works
+        key = metatype,
+        include_when = "inode/directory"
+			# The grouper will not only create groups but can provide additional
+			# information about the groups. This additional information will be
+			# available with the names `current_metatype` and `is_directory`.
+      # If the `With` clause is omitted, teh grouper will use default names
```
Comment (iainrb) [inline language.md]: "teh" -> "the"

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/Grouper.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * Perform a complex sub grouping operation
+ *
+ * @param <I> the input record type
+ * @param <O> the output record type
+ */
+public interface Grouper<I, O> {
+  /**
+   * Group the inputs provided into sensible subgroups. The grouper may drop input or assign the
+   * same input to multiple groups; it should not include the same input the same group multiple
```
Comment (iainrb) [inline shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/Grouper.java]: "same input the same group" -> "same input in the same group"

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/Grouper.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+   * times
+   *
+   * @param inputs the inputs to be grouped; the list is now owned by the grouper and the grouper
```
Comment (iainrb) [inline shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/Grouper.java]: "inputs the inputs" "the inputs"

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/GrouperDefinition.java [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+                return factory.pack(
+                    parameter1.unpack(first), parameter2.unpack(second), outputGenerator);
+              }
+            }));
+    this.inputs = Arrays.asList(parameter1, parameter2);
+    this.outputs = outputs;
+  }
+
+  /**
+   * Define a new grouper that takes two arguments
```
Comment (iainrb) [inline shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/GrouperDefinition.java]: "two" -> "three"

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/Grouper.java (hunk 10 above)
Comment (apmasell) [reply shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/Grouper.java]: The first `inputs` is part of the JavaDoc identifier. It doesn't highlight properly here.

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/GrouperDefinition.java [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  /**
+   * Define a new grouper that takes three arguments
+   *
+   * @param name the name of the grouper that should be used in the olive; it must be a valid Shesmu
+   *     identifier
+   * @param parameter1 the first input the olives must provide
+   * @param parameter2 the second input the olives must provide
+   * @param outputs the collector format the olive will provided for exporting the extra subgroup
```
Comment (iainrb) [inline shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/grouper/GrouperDefinition.java]: @param parameter3 the third input the olives must provide

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/types/GenericReturnTypeGuarantee.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+   * Create a type variable
+   *
+   * @param clazz the upper bound for the type variable
+   * @param id a unique name for the variable; if it is not unique in the context of inference, bad
+   *     things will happen
+   * @param <T> the Java type for the Shesmu type being transited through Java
+   */
+  public static <T> Pair<GenericTypeGuarantee<T>, GenericReturnTypeGuarantee<T>> variable(
+      Class<? extends T> clazz, String id) {
+    // This is a situation that require multiple inheirtance, so what we do is produce two values
```
Comment (iainrb) [inline shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/types/GenericReturnTypeGuarantee.java]: "inheirtance" -> "inheritance"

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/types/GenericTypeGuarantee.java [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+    };
+  }
+
+  /**
+   * Create a tuple that contains generic types
+   *
+   * @param pack a function to convert the tuple's value to the preferred Java type
+   * @param first the type of the first element in the tuple
+   * @param second the type of the second element in the tuple
+   * @param <R> the type of value after convertion to a Java type
```
Comment (iainrb) [inline shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/types/GenericTypeGuarantee.java]: "convertion" -> "conversion"

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/BaseOliveBuilder.java [hunk 15]
```diff
@@ -364,6 +373,167 @@ public final void measureFlow(String filename, int line, int column) {
...
+          LambdaBuilder.pushVirtual(
+              renderer,
+              RegroupVariablesBuilder.METHOD_IS_OK.getName(),
+              LambdaBuilder.predicate(newType));
+          renderer.methodGen().invokeInterface(A_STREAM_TYPE, METHOD_STREAM__FILTER);
+        });
+
+    final Renderer newRenderer = newLambda.renderer(oldType, this::emitSigner);
+    final Renderer collectedRenderer = collectLambda.renderer(oldType, 1, this::emitSigner);
+    // For any grouper variable that is really a function, redefine it to be a the correct function
```
Comment (iainrb) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/BaseOliveBuilder.java]: " to be a the correct function call" -> " to be the correct function call"

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/OliveClauseNodeGroupWithGrouper.java [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+                .toArray(LoadableValue[]::new));
+
+    discriminators.forEach(d -> d.render(regroup));
+    children.forEach(group -> group.render(regroup, builder));
+    regroup.finish();
+
+    oliveBuilder.measureFlow(builder.sourcePath(), line, column);
+    // We have now thoroughly hosed equals and hashCode in the output. If the next clause was
+    // another grouper and it tried to stick things in a hashmap, everything might be endless
+    // suffering. So, we make a hidden 1:1 let clause to stop the chaos.
```
Comment (iainrb) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/OliveClauseNodeGroupWithGrouper.java]: 👍

**Pull request merged**